In [9]:
sample_list = [1, 2, 3, 4, 5]

for idx, element in enumerate(sample_list):
    element +=1
    sample_list[idx] = element

print(sample_list)

[2, 3, 4, 5, 6]


In [1]:
import ssl

ssl._create_default_https_context = ssl._create_unverified_context

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.tree import plot_tree

from sklearn.model_selection import train_test_split
# Regression
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_absolute_error, r2_score, mean_squared_error
# Clasification
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, precision_score, accuracy_score
# Best_searches
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
# Standarization
from sklearn.preprocessing import StandardScaler

from keras import models
from keras.layers import Dense

In [2]:
df_train = pd.read_csv('https://raw.githubusercontent.com/Mateusz-Matejko/kaggle/main/projects/competition-clasification-binary-of-machine-failures/data/train.csv', encoding='utf8')
df_test = pd.read_csv('https://raw.githubusercontent.com/Mateusz-Matejko/kaggle/main/projects/competition-clasification-binary-of-machine-failures/data/test.csv', encoding='utf8')
df_final = pd.read_csv('https://raw.githubusercontent.com/Mateusz-Matejko/kaggle/main/projects/competition-clasification-binary-of-machine-failures/data/sample_submission.csv', encoding='utf8')

In [3]:
df_train.describe()

id  Air temperature [K]  Process temperature [K]   
count  136429.000000        136429.000000            136429.000000  \
mean    68214.000000           299.862776               309.941070   
std     39383.804275             1.862247                 1.385173   
min         0.000000           295.300000               305.800000   
25%     34107.000000           298.300000               308.700000   
50%     68214.000000           300.000000               310.000000   
75%    102321.000000           301.200000               310.900000   
max    136428.000000           304.400000               313.800000   

       Rotational speed [rpm]    Torque [Nm]  Tool wear [min]   
count           136429.000000  136429.000000    136429.000000  \
mean              1520.331110      40.348643       104.408901   
std                138.736632       8.502229        63.965040   
min               1181.000000       3.800000         0.000000   
25%               1432.000000      34.600000        48.000000   
50%               1493.000000      40.400000       106.000000   
75%               1580.000000      46.100000       159.000000   
max               2886.000000      76.600000       253.000000   

       Machine failure            TWF            HDF            PWF   
count    136429.000000  136429.000000  136429.000000  136429.000000  \
mean          0.015744       0.001554       0.005160       0.002397   
std           0.124486       0.039389       0.071649       0.048899   
min           0.000000       0.000000       0.000000       0.000000   
25%           0.000000       0.000000       0.000000       0.000000   
50%           0.000000       0.000000       0.000000       0.000000   
75%           0.000000       0.000000       0.000000       0.000000   
max           1.000000       1.000000       1.000000       1.000000   

                 OSF            RNF  
count  136429.000000  136429.000000  
mean        0.003958       0.002258  
std         0.062789       0.047461  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%         0.000000       0.000000  
max         1.000000       1.000000

In [4]:
df_test[df_test['PWF'] != 0]

id Product ID Type  Air temperature [K]  Process temperature [K]   
405    136834     L47380    L                298.4                    308.6  \
576    137005     L56355    L                297.9                    309.0   
950    137379     L49935    L                299.7                    309.3   
1570   137999     L55609    L                299.0                    310.4   
2653   139082     L53074    L                301.2                    311.0   
...       ...        ...  ...                  ...                      ...   
86883  223312     M20301    M                303.0                    312.6   
87045  223474     L49512    L                299.2                    308.5   
89247  225676     M18633    M                302.4                    311.1   
90239  226668     L51961    L                303.4                    311.7   
90665  227094     M18884    M                302.1                    311.0   

       Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  TWF  HDF  PWF   
405                      2636         12.8              103    0    0    1  \
576                      1403         64.2                0    0    0    1   
950                      1478         38.8                8    0    0    1   
1570                     2861          4.6               92    0    0    1   
2653                     1267         69.3              164    0    0    1   
...                       ...          ...              ...  ...  ...  ...   
86883                    1363         62.9              119    0    0    1   
87045                    2232         15.9              220    0    0    1   
89247                    2659         10.8              139    0    0    1   
90239                    1374         63.7              121    0    0    1   
90665                    1479         60.3               31    0    0    1   

       OSF  RNF  
405      0    0  
576      0    0  
950      0    0  
1570     0    0  
2653     0    0  
...    ...  ...  
86883    0    0  
87045    0    0  
89247    0    0  
90239    0    0  
90665    0    0  

[214 rows x 13 columns]

In [5]:
df_train.shape

(136429, 14)

In [6]:
df_train[df_train['Machine failure'] != 0].shape

(2148, 14)

## Opis danych: 

id - 
    Wiemy o co chodzi
    
Podcut ID - 
    Pomysły: Zmapować konkretny produkt id, jako taki z dużą lub małą szansą na uszkodzenie. 
    
### Zrobione 
Type - ( H, M, L )(High, Medium, Low) - 
    Im wyższy tym mniejsza szansa na uszkodzenie i większa tolerancja na długotrwałą pracę


Air temparture - temperatura powietrza
    zakres: 295 - 304 stopnie, median 300
    Różnica między tą temperaturą a temperaturą procesu może wskazywać na przegrzewanie.
    
### Zrobione
<i>Pomysły: Zakodować różnicę między temperaturami.</i>
    
Process temperature - temperatura pracy
    zakres 305 - 314, median 310.
    Pomysły: Sprawdzić zależność między awariami a tą temperaturą, DLA WSZYSTKICH TYPU AWARII

### Zrobione 
<i> Note: ZAPYTAĆ KACPRA CZY WARTO SPRAWDZIĆ TO NA WARUNKU + 0.5C </i>
<i>Jeżeli Air Temprature będzie równe Process temp lub większe, użądzenie nie będzie się w stanie schłodzić i się nagrzeje.</i> 

### In progress
Rotational speed rpm - prędkość rotacji/obrotu
    zakres 1181 - 2886, 996 unikalnych
    Pomysł: Porównać z Tool Wear 
    
Torque Nm - Moment obrotowy 
    Pomysł: użyć wzór, spróbować oszacować błąd. Power (kW) = Torque (N.m) x Speed (RPM) z TWF
    Możliwe: powiązanie OSF
    
Tool wear - Czas sesji użycia narzędzia
    Zakres 0 - 253
    Pomysł: Im większy tool wear, może powodować awarie
    
Typy awarii:
    TWF - Zużycie
    HDF - Przegrzanie
    PWF - Za mała moc
    OSF - Przeciążenie mechaniczne
    RNF - Losowe
    Pomysły: Zobaczyć zależności pomiędzy tymi typami awarii i dokładnie je zbadać.
    
Zbadać: outliersy dlaczego min czas sesji = 0

In [7]:
df_train.describe(include='all')

id Product ID    Type  Air temperature [K]   
count   136429.000000     136429  136429        136429.000000  \
unique            NaN       9976       3                  NaN   
top               NaN     L53257       L                  NaN   
freq              NaN        139   95354                  NaN   
mean     68214.000000        NaN     NaN           299.862776   
std      39383.804275        NaN     NaN             1.862247   
min          0.000000        NaN     NaN           295.300000   
25%      34107.000000        NaN     NaN           298.300000   
50%      68214.000000        NaN     NaN           300.000000   
75%     102321.000000        NaN     NaN           301.200000   
max     136428.000000        NaN     NaN           304.400000   

        Process temperature [K]  Rotational speed [rpm]    Torque [Nm]   
count             136429.000000           136429.000000  136429.000000  \
unique                      NaN                     NaN            NaN   
top                         NaN                     NaN            NaN   
freq                        NaN                     NaN            NaN   
mean                 309.941070             1520.331110      40.348643   
std                    1.385173              138.736632       8.502229   
min                  305.800000             1181.000000       3.800000   
25%                  308.700000             1432.000000      34.600000   
50%                  310.000000             1493.000000      40.400000   
75%                  310.900000             1580.000000      46.100000   
max                  313.800000             2886.000000      76.600000   

        Tool wear [min]  Machine failure            TWF            HDF   
count     136429.000000    136429.000000  136429.000000  136429.000000  \
unique              NaN              NaN            NaN            NaN   
top                 NaN              NaN            NaN            NaN   
freq                NaN              NaN            NaN            NaN   
mean         104.408901         0.015744       0.001554       0.005160   
std           63.965040         0.124486       0.039389       0.071649   
min            0.000000         0.000000       0.000000       0.000000   
25%           48.000000         0.000000       0.000000       0.000000   
50%          106.000000         0.000000       0.000000       0.000000   
75%          159.000000         0.000000       0.000000       0.000000   
max          253.000000         1.000000       1.000000       1.000000   

                  PWF            OSF            RNF  
count   136429.000000  136429.000000  136429.000000  
unique            NaN            NaN            NaN  
top               NaN            NaN            NaN  
freq              NaN            NaN            NaN  
mean         0.002397       0.003958       0.002258  
std          0.048899       0.062789       0.047461  
min          0.000000       0.000000       0.000000  
25%          0.000000       0.000000       0.000000  
50%          0.000000       0.000000       0.000000  
75%          0.000000       0.000000       0.000000  
max          1.000000       1.000000       1.000000

In [16]:
df_train.isnull().sum().sum()

0

### Power (kW) = Torque (N.m) x Speed (RPM)

## Get to know your data

In [9]:
df_train['Product ID'].unique().shape #(9976,)
df_train['Type'].unique() #array(['L', 'M', 'H'], dtype=object)
df_train['Air temperature [K]'] #(95,)
df_train['Process temperature [K]'].unique().shape #(81,)
df_train['Tool wear [min]'].unique().shape #(246,)
df_train['Rotational speed [rpm]'].unique().shape #(952,)
df_train['Torque [Nm]'].unique().shape #(611,)
df_train['Tool wear [min]'].unique().shape #(246,)

(246,)

In [10]:
# Dataset shape
print(df_train.shape, "\n") # (74051, 10)

# Find out any Null values
df_train.isnull().sum() # 0 null values

(136429, 14) 



id                         0
Product ID                 0
Type                       0
Air temperature [K]        0
Process temperature [K]    0
Rotational speed [rpm]     0
Torque [Nm]                0
Tool wear [min]            0
Machine failure            0
TWF                        0
HDF                        0
PWF                        0
OSF                        0
RNF                        0
dtype: int64

In [11]:
df_train.head()

id Product ID Type  Air temperature [K]  Process temperature [K]   
0   0     L50096    L                300.6                    309.6  \
1   1     M20343    M                302.6                    312.1   
2   2     L49454    L                299.3                    308.5   
3   3     L53355    L                301.0                    310.9   
4   4     M24050    M                298.0                    309.0   

   Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  Machine failure  TWF   
0                    1596         36.1              140                0    0  \
1                    1759         29.1              200                0    0   
2                    1805         26.5               25                0    0   
3                    1524         44.3              197                0    0   
4                    1641         35.4               34                0    0   

   HDF  PWF  OSF  RNF  
0    0    0    0    0  
1    0    0    0    0  
2    0    0    0    0  
3    0    0    0    0  
4    0    0    0    0

In [12]:
df_train.groupby('Type')['Machine failure'].mean()

Type
H    0.013000
L    0.016727
M    0.013592
Name: Machine failure, dtype: float64

In [13]:
type_dict = {
    'H': 0.013000,
    'M': 0.013592,
    'L': 0.016727
}

<i>Jeżeli Air Temprature będzie równe Process temp lub większe, użądzenie nie będzie się w stanie schłodzić i się nagrzeje.</i> 

1. Potrzebne dane:
    - Air temprature = 6
    - Process temperature = 4
2. Porównanie:
    - 

In [81]:
df_train = pd.read_csv('https://raw.githubusercontent.com/Mateusz-Matejko/kaggle/main/projects/competition-clasification-binary-of-machine-failures/data/train.csv', encoding='utf8')

def process_data(df:pd.DataFrame):
    df['Power [KW]'] = df['Torque [Nm]'] * df['Rotational speed [rpm]']
    df['Type'] = df['Type'].map(type_dict)
    df['Air temperature [C]'] = df['Air temperature [K]'] - 272.15
    df['Process temperature [C]'] = df['Process temperature [K]'] - 272.15
    df = df.drop(labels=['Air temperature [K]', 'Process temperature [K]'], axis=1)
    print(df.columns)
    df['Possbile_overheating'] = (df['Air temperature [C]'] >= df['Process temperature [C]']) #possibly add sth to C value
    df['Temperature difference'] = df['Process temperature [C]'] - df['Air temperature [C]']

    return df

In [82]:
df_train=process_data(df_train)

Index(['id', 'Product ID', 'Type', 'Rotational speed [rpm]', 'Torque [Nm]',
       'Tool wear [min]', 'Machine failure', 'TWF', 'HDF', 'PWF', 'OSF', 'RNF',
       'Power [KW]', 'Air temperature [C]', 'Process temperature [C]'],
      dtype='object')


In [83]:
df_train.head()

id Product ID      Type  Rotational speed [rpm]  Torque [Nm]   
0   0     L50096  0.016727                    1596         36.1  \
1   1     M20343  0.013592                    1759         29.1   
2   2     L49454  0.016727                    1805         26.5   
3   3     L53355  0.016727                    1524         44.3   
4   4     M24050  0.013592                    1641         35.4   

   Tool wear [min]  Machine failure  TWF  HDF  PWF  OSF  RNF  Power [KW]   
0              140                0    0    0    0    0    0     57615.6  \
1              200                0    0    0    0    0    0     51186.9   
2               25                0    0    0    0    0    0     47832.5   
3              197                0    0    0    0    0    0     67513.2   
4               34                0    0    0    0    0    0     58091.4   

   Air temperature [C]  Process temperature [C]  Possbile_overheating   
0                28.45                    37.45                 False  \
1                30.45                    39.95                 False   
2                27.15                    36.35                 False   
3                28.85                    38.75                 False   
4                25.85                    36.85                 False   

   Temperature difference  
0                     9.0  
1                     9.5  
2                     9.2  
3                     9.9  
4                    11.0

In [106]:
pd.DataFrame(index=[df_train.groupby('Tool wear [min]')['Machine failure']])

Empty DataFrame
Columns: []
Index: [((0, 18        0
27        0
91        0
92        0
93        0
         ..
136285    0
136302    0
136348    0
136361    0
136379    0
Name: Machine failure, Length: 3790, dtype: int64), (2, 26        0
153       0
187       0
191       0
276       0
         ..
136175    0
136228    0
136281    0
136363    0
136425    0
Name: Machine failure, Length: 1771, dtype: int64), (3, 181       0
288       0
1486      0
1577      0
2477      0
         ..
133514    0
135611    0
135975    0
136156    0
136209    0
Name: Machine failure, Length: 255, dtype: int64), (4, 621       0
918       0
1798      0
1932      0
2048      0
         ..
134252    0
134558    0
134885    0
134963    0
136136    0
Name: Machine failure, Length: 313, dtype: int64), (5, 17        0
186       0
285       0
337       0
351       0
         ..
135489    0
135595    0
136031    0
136195    0
136329    0
Name: Machine failure, Length: 1419, dtype: int64), (6, 238       0
270       0
458       0
713       0
3121      0
         ..
134105    1
134257    0
135132    0
135617    0
136081    0
Name: Machine failure, Length: 244, dtype: int64), (7, 429       0
495       0
514       0
562       0
596       0
         ..
136089    0
136137    0
136259    0
136291    0
136318    0
Name: Machine failure, Length: 1342, dtype: int64), (8, 711       0
1176      0
1713      0
1903      0
2439      0
         ..
132245    0
132586    0
134808    0
135179    0
136401    0
Name: Machine failure, Length: 337, dtype: int64), (9, 46        0
57        0
198       0
294       0
373       0
         ..
136024    0
136090    0
136125    0
136145    0
136390    0
Name: Machine failure, Length: 1329, dtype: int64), (10, 105       0
694       0
787       0
1117      0
1203      0
         ..
132747    0
133204    0
133458    0
133930    0
134159    0
Name: Machine failure, Length: 492, dtype: int64), (11, 524       0
567       0
643       0
1910      0
2166      0
         ..
135041    0
135189    0
135685    0
135864    0
135910    0
Name: Machine failure, Length: 439, dtype: int64), (12, 251       0
264       0
293       0
744       0
806       0
         ..
135881    0
136072    0
136116    0
136142    0
136180    0
Name: Machine failure, Length: 785, dtype: int64), (13, 8         0
73        0
265       0
399       0
746       0
         ..
135967    0
136016    0
136057    0
136230    0
136367    0
Name: Machine failure, Length: 822, dtype: int64), (14, 43        0
434       0
499       0
638       0
1213      0
         ..
134835    0
135000    0
135944    0
136115    0
136403    0
Name: Machine failure, Length: 518, dtype: int64), (15, 163       0
214       0
252       0
406       0
527       0
         ..
135830    0
135912    0
135926    0
135978    0
136108    0
Name: Machine failure, Length: 1036, dtype: int64), (16, 124       0
281       0
1081      0
1325      0
1369      0
         ..
135646    0
135834    0
135841    0
136050    0
136097    0
Name: Machine failure, Length: 667, dtype: int64), (17, 171       0
382       0
389       0
504       0
529       0
         ..
135840    0
135849    0
135895    0
135970    0
136295    0
Name: Machine failure, Length: 1102, dtype: int64), (18, 121       0
150       0
221       0
559       0
636       0
         ..
134990    0
135470    0
135940    0
136199    0
136421    0
Name: Machine failure, Length: 699, dtype: int64), (19, 36        0
693       0
695       0
719       0
864       0
         ..
134804    0
134929    0
135299    0
135651    0
136060    0
Name: Machine failure, Length: 448, dtype: int64), (20, 49        0
131       0
228       1
578       0
697       0
         ..
135680    0
135858    0
135914    0
135961    0
136048    0
Name: Machine failure, Length: 772, dtype: int64), (21, 130       1
225       0
341       0
465       0
606       0
         ..
134919    0
135200    0
135326    0
135726    0
135796    0
Name: Machine failure, Length: 532, dtype: int64), (22, 209       0


In [ ]:
pd.DataFrame(index=[df_train.groupby('Tool wear [min]')['Machine failure'].mean(),
                   df_train['Product ID'].unique()])

In [120]:
pd.DataFrame(index=[df_train.groupby('Tool wear [min]')['Machine failure'].mean(),
                   df_train['Product ID'].unique()])

ValueError: all arrays must be same length

In [24]:
df_train.head()

id Product ID      Type  Rotational speed [rpm]  Torque [Nm]   
0   0     L50096  0.016727                    1596         36.1  \
1   1     M20343  0.013592                    1759         29.1   
2   2     L49454  0.016727                    1805         26.5   
3   3     L53355  0.016727                    1524         44.3   
4   4     M24050  0.013592                    1641         35.4   

   Tool wear [min]  Machine failure  TWF  HDF  PWF  OSF  RNF  Power [KW]   
0              140                0    0    0    0    0    0     57615.6  \
1              200                0    0    0    0    0    0     51186.9   
2               25                0    0    0    0    0    0     47832.5   
3              197                0    0    0    0    0    0     67513.2   
4               34                0    0    0    0    0    0     58091.4   

   Air temperature [C]  Process temperature [C]  Possbile_overheating   
0                28.45                    37.45                  True  \
1                30.45                    39.95                  True   
2                27.15                    36.35                  True   
3                28.85                    38.75                  True   
4                25.85                    36.85                  True   

   Temperature difference  
0                    -9.0  
1                    -9.5  
2                    -9.2  
3                    -9.9  
4                   -11.0

##Pochwalić się Kacprowi ;3


In [48]:
pd.DataFrame(df_train.groupby('Tool wear [min]')['Machine failure'].mean()).describe()

Machine failure
count       246.000000
mean          0.034653
std           0.085217
min           0.000000
25%           0.006726
50%           0.011525
75%           0.023674
max           1.000000

In [49]:
pd.DataFrame(df_train.groupby('Rotational speed [rpm]')['Machine failure'].mean()).describe()

Machine failure
count       952.000000
mean          0.056718
std           0.168048
min           0.000000
25%           0.000000
50%           0.000000
75%           0.015444
max           1.000000